delta/sdk/src/main/scala/ch/epfl/bluebrain/nexus/delta/sdk/Resources.scala

```diff
@@ -0,0 +1,305 @@
+    * @return a paginated results list of resources sorted by their creation date.
+    */
+  def list(
+      pagination: FromPagination,
+      params: ResourceSearchParams = ResourceSearchParams.none
+  ): UIO[UnscoredSearchResults[DataResource]]
+
+  /**
+    * A non terminating stream of events for resources. After emitting all known events it sleeps until new events
+    * are recorded.
+    *
+    * @param offset the last seen event offset; it will not be emitted by the stream
+    */
+  def events(offset: Offset = NoOffset): Stream[Task, Envelope[ResourceEvent]]
+
+  /**
+    * The current resource events. The stream stops after emitting all known events.
+    *
+    * @param offset the last seen event offset; it will not be emitted by the stream
+    */
+  def currentEvents(offset: Offset = NoOffset): Stream[Task, Envelope[ResourceEvent]]
+
+}
+
+object Resources {
+
+  /**
+    * The resources module type.
+    */
+  final val moduleType: String = "resource"
+
+  private[delta] def next(state: ResourceState, event: ResourceEvent): ResourceState = {
+    // format: off
+    def created(e: ResourceCreated): ResourceState = state match {
+      case Initial     => Current(e.id, e.project, e.source, e.compacted, e.expanded, e.rev, deprecated = false, e.schema, e.types, Map.empty, e.instant, e.subject, e.instant, e.subject)
+      case s: Current  => s
+    }
+
+    def updated(e: ResourceUpdated): ResourceState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, types = e.types, source = e.source, compacted = e.compacted, expanded = e.expanded, updatedAt = e.instant, updatedBy = e.subject)
+    }
+
+    def tagAdded(e: ResourceTagAdded): ResourceState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, tags = s.tags + (e.tag -> e.targetRev), updatedAt = e.instant, updatedBy = e.subject)
+    }
+    // format: on
+
+    def deprecated(e: ResourceDeprecated): ResourceState = state match {
+      case Initial    => Initial
+      case s: Current => s.copy(rev = e.rev, deprecated = true, updatedAt = e.instant, updatedBy = e.subject)
+    }
+    event match {
+      case e: ResourceCreated    => created(e)
+      case e: ResourceUpdated    => updated(e)
+      case e: ResourceTagAdded   => tagAdded(e)
+      case e: ResourceDeprecated => deprecated(e)
+    }
+  }
+
+  @SuppressWarnings(Array("OptionGet"))
+  private[delta] def evaluate(
+      fetchSchema: ResourceRef => UIO[Option[SchemaResource]]
+  )(state: ResourceState, cmd: ResourceCommand)(implicit
+      clock: Clock[UIO] = IO.clock,
+      rcr: RemoteContextResolution
+  ): IO[ResourceRejection, ResourceEvent] = {
+
+    def asGraph(id: Iri, expanded: ExpandedJsonLd): IO[ResourceRejection, Graph] =
+      expanded.toGraph.leftMap(err => ResourceJsonLdPayloadRejection(Some(id), err))
+
+    def validate(schemaRef: ResourceRef, id: Iri, graph: Graph): IO[ResourceRejection, Unit] = {
```

Comment:
I think the entire validation logic should be the "schema's" responsibility. It would simplify this logic if the STM would only take in a validation function `Graph => UIO[InvalidResource, Unit]`.
This would also enable for later schema caching...

Reply:
We will see when schemas is done, then the logic can be ported if that's what we want. Now it fits well here, since resources are the only thing that needs to be validated by schemas. So it is quite self contained.

Reply:
Happy to consider changing this in a subsequent PR when schemas are implemented.